internal/core/server/kernel.go

```diff
@@ -28,6 +30,45 @@ type Kernel struct {
 	TaskTracker *task.TaskTracker
 }
 
+func (kern *Kernel) DescribeTemplates(ctx context.Context, input *api.DescribeTemplatesInput) (*api.DescribeTemplatesOutput, error) {
+	return &api.DescribeTemplatesOutput{Templates: template.GetTemplateDescriptions()}, nil
+}
+
+func (kern *Kernel) CreateProject(ctx context.Context, input *api.CreateProjectInput) (*api.CreateProjectOutput, error) {
+	projectDir := input.Root
+	if !filepath.IsAbs(projectDir) {
+		return &api.CreateProjectOutput{}, errors.New("path must be absolute")
+	}
+
+	var templateDir string
+	if input.TemplateUrl != nil {
```

Comment:
Empty string is an invalid URL, so might as well enbrace the zero-value idiom and change the type to just be string instead of `*string` then check against `""` here.

Reply:
From the Go perspective that seems reasonable — but do we want that to be true at the API interface? That feels like leaking the implementation detail of the language. Nbd either way.

Reply:
I'm doing some work on formalizing JOSH & going to replace the IDL format with something JSON-based & try to define the schema such that we can generate sensible types for both Go and TypeScript. Stay tuned for that. Very early work on it here: https://github.com/deref/josh